### PR TITLE
Split cilium binary from cilium-agent

### PR DESCRIPTION
**Pull request by aanm**
Although having cilium built in the cilium-agent binary makes the toal
size of both binaries smaller than having them split it also causes the
memory consumption of the cilium-agent to be larger than necessary.

Having 1 single binary results in a 100MB file vs 95+57=152MB when
spliting into 2 binaries. However, the memory consumption of cilium-agent
decreases from 82.85MB to 73.93MB by not importing all of the cilium/cmd
code. The following changes will decreases the size of the cilium client
so the total size will not be as big as 152MB.

Unfortunately there is not much we can do WRT the size of the cilium binary as there are functionalities used in the `cilium policy trace` that depends in kubernetes packages. e.g.:
```golang
		// Parse pod names.
		if srcK8sPod != "" {
			id, err := getSecIDFromK8s(srcK8sPod)
			if err != nil {
				Fatalf("Cannot get security id from k8s pod name: %s", err)
			}
			srcSlice = appendIdentityLabelsToSlice(srcSlice, id)
			srcSlices = append(srcSlices, srcSlice)
		}

		if dstK8sPod != "" {
			id, err := getSecIDFromK8s(dstK8sPod)
			if err != nil {
				Fatalf("Cannot get security id from k8s pod name: %s", err)
			}
			dstSlice = appendIdentityLabelsToSlice(dstSlice, id)
			dstSlices = append(dstSlices, dstSlice)
		}

		// Parse provided YAML files.
		if srcK8sYaml != "" {
			srcYamlSlices, err := trace.GetLabelsFromYaml(srcK8sYaml)
			if err != nil {
				Fatalf("%s", err)
			}
			for _, v := range srcYamlSlices {
				srcSlices = append(srcSlices, v)
			}
		}

		if dstK8sYaml != "" {
			dstYamlSlices, err := trace.GetLabelsFromYaml(dstK8sYaml)
			if err != nil {
				Fatalf("%s", err)
			}
			for _, v := range dstYamlSlices {
				dstSlices = append(dstSlices, v)
			}
		}
```

For now, `strip` is the only tool used to reduce the size of the cilium binary.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/8281)

### CODE REVIEW

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by nebril** [comment]
hit by #8305

**Comment by nebril** [comment]
test-me-please

**Comment by borkmann** [comment]
> Although having cilium built in the cilium-agent binary makes the toal
> size of both binaries smaller than having them split it also causes the
> memory consumption of the cilium-agent to be larger than necessary.
> 
> Having 1 single binary results in a 100MB file vs 95+57=152MB when
> spliting into 2 binaries. However, the memory consumption of cilium-agent
> decreases from 82.85MB to 73.93MB by not importing all of the cilium/cmd
> code. The following changes will decreases the size of the cilium client
> so the total size will not be as big as 152MB.

Given this circumstance of 52M increase for just 10M memory consumption decrease, is this split really worth it? Feels kind of borderline, no?

**Comment by aanm** [comment]
> > Although having cilium built in the cilium-agent binary makes the toal
> > size of both binaries smaller than having them split it also causes the
> > memory consumption of the cilium-agent to be larger than necessary.
> > Having 1 single binary results in a 100MB file vs 95+57=152MB when
> > spliting into 2 binaries. However, the memory consumption of cilium-agent
> > decreases from 82.85MB to 73.93MB by not importing all of the cilium/cmd
> > code. The following changes will decreases the size of the cilium client
> > so the total size will not be as big as 152MB.
> 
> Given this circumstance of 52M increase for just 10M memory consumption decrease, is this split really worth it? Feels kind of borderline, no?

@borkmann the last commit changes the filesize of the cilium client to `42M` so it's 37M vs 10M. It's not only memory consumption but there are also init() functions that are uselessly executed in the cilium-agent because a) those init functions are automatically started when the package is imported b) even if the agent does not use that package directly, those functions will start and will run in the background.

**Comment by brb** [comment]
Another thing to consider is that the Cilium Docker image increases in its size.

File: cilium/Makefile [hunk 1]
```diff
@@ -7,6 +7,7 @@ SOURCES := $(shell find ../api ../daemon ../common ../pkg cmd . \( -name '*.go'
 $(TARGET): $(SOURCES)
 	@$(ECHO_GO)
 	$(QUIET)$(GO) build $(GOBUILD) -o $(TARGET)
+	strip $(TARGET)
```
Comment (raybejjani) [inline cilium/Makefile:10]: Doesn't this break panic backtraces? does it break anything at all?

Comment (aanm) [reply cilium/Makefile:10]: @raybejjani do we care about panics in the client?

**Review by tgraf — APPROVED** [review]

**Pull request merged**